### PR TITLE
Reduce access levels on classes and fields in DPS service client

### PR DESCRIPTION
**Pull request by timtay-microsoft**
Also convert handwritten getters and setters to lombok

### CODE REVIEW

File: provisioning/provisioning-service-client/src/main/java/com/microsoft/azure/sdk/iot/provisioning/service/EnrollmentGroupManager.java [hunk 1]
```diff
@@ -25,7 +25,7 @@
  * @see <a href="https://docs.microsoft.com/en-us/azure/iot-dps/">Azure IoT Hub Device Provisioning Service</a>
  * @see <a href="https://docs.microsoft.com/en-us/rest/api/iot-dps/deviceenrollmentgroup">Device Enrollment Group</a>
  */
-public class EnrollmentGroupManager
+class EnrollmentGroupManager
```
Comment (timtay-microsoft) [inline provisioning/provisioning-service-client/src/main/java/com/microsoft/azure/sdk/iot/provisioning/service/EnrollmentGroupManager.java:28]: This class and other "Manager" class are just 1 level below our public client, but they shouldn't directly be used

File: provisioning/provisioning-service-client/src/main/java/com/microsoft/azure/sdk/iot/provisioning/service/Tools.java [hunk 2]
```diff
@@ -248,23 +248,4 @@ public static void appendJsonAttribute(StringBuilder strBuilder, String name, St
...
     }
-
-    /**
-     * Equality check for objects that accounts for null value comparisons. If both objects are null, this will return false.
-     * Both objects must have .equals(...) implemented correctly for this method to work properly.
-     * @param a the first object
-     * @param b the seconds object
-     * @return if the two are equal
-     */
-    public static boolean areEqual(Object a, Object b)
```
Comment (timtay-microsoft) [inline provisioning/provisioning-service-client/src/main/java/com/microsoft/azure/sdk/iot/provisioning/service/Tools.java:259]: unused

File: provisioning/provisioning-service-client/src/main/java/com/microsoft/azure/sdk/iot/provisioning/service/configs/DeviceRegistrationState.java [hunk 3]
```diff
@@ -41,59 +42,68 @@
...
+    @Getter
+    private transient Date createdDateTimeUtc;
 
     // last update date and time
     private static final String LAST_UPDATED_DATETIME_UTC_TAG = "lastUpdatedDateTimeUtc";
     @Expose
     @SerializedName(LAST_UPDATED_DATETIME_UTC_TAG)
     @SuppressWarnings("unused") // used by reflection during json serialization/deserialization
-    private String lastUpdatedDateTimeUtc;
-    private transient Date lastUpdatedDateTimeUtcDate;
```
Comment (timtay-microsoft) [inline provisioning/provisioning-service-client/src/main/java/com/microsoft/azure/sdk/iot/provisioning/service/configs/DeviceRegistrationState.java:66]: Swapping the names of these fields a bit to maintain that deviceRegistrationState.getLastUpdatedDateTimeUtc() still returns a date rather than a string

Comment (timtay-microsoft) [reply provisioning/provisioning-service-client/src/main/java/com/microsoft/azure/sdk/iot/provisioning/service/configs/DeviceRegistrationState.java:66]: You'll see this same pattern in a few date time fields across a few serialization types

File: provisioning/provisioning-service-client/src/test/java/com/microsoft/azure/sdk/iot/provisioning/service/configs/TpmAttestationTest.java [hunk 4]
```diff
@@ -18,28 +18,6 @@
     private static final String VALID_ENDORSEMENT_KEY = "validEndorsementKey";
     private static final String VALID_STORAGE_ROOT_KEY = "validStorageRootKey";
 
-    /* SRS_TPM_ATTESTATION_21_001: [The constructor shall throw IllegalArgumentException if the provided endorsementKey is null or empty.] */
```
Comment (timtay-microsoft) [inline provisioning/provisioning-service-client/src/test/java/com/microsoft/azure/sdk/iot/provisioning/service/configs/TpmAttestationTest.java:21]: I removed some unnecessary null/empty checks, so a lot of unit tests needed to be deleted

File: provisioning/provisioning-service-client/src/main/java/com/microsoft/azure/sdk/iot/provisioning/service/Query.java [hunk 5]
```diff
@@ -101,8 +100,7 @@
      * @param pageSize the {@code int} with the maximum number of items per iteration. It cannot be negative.
      * @throws IllegalArgumentException if one of the parameters is invalid.
      */
-    @SuppressWarnings("deprecation")
-    protected Query(ContractApiHttp contractApiHttp, String targetPath, QuerySpecification querySpecification, int pageSize)
+    Query(ContractApiHttp contractApiHttp, String targetPath, QuerySpecification querySpecification, int pageSize)
```
Comment (abhipsaMisra) [inline provisioning/provisioning-service-client/src/main/java/com/microsoft/azure/sdk/iot/provisioning/service/Query.java:103]: what about the "deprecation"?

File: provisioning/provisioning-service-client/src/main/java/com/microsoft/azure/sdk/iot/provisioning/service/auth/ProvisioningConnectionString.java [hunk 6]
```diff
@@ -3,32 +3,53 @@
...
 
     // Included in the device connection string
-    protected String hostName;
-    protected String deviceProvisioningServiceName;
-    protected AuthenticationMethod authenticationMethod;
-    protected String sharedAccessKeyName;
-    protected String sharedAccessKey;
-    protected String sharedAccessSignature;
+    @Getter
+    @Setter(AccessLevel.MODULE)
```
Comment (abhipsaMisra) [inline provisioning/provisioning-service-client/src/main/java/com/microsoft/azure/sdk/iot/provisioning/service/auth/ProvisioningConnectionString.java:31]: q - is there a reason you chose module instead of package?
looks like they mean the same thing: https://github.com/projectlombok/lombok/blob/e050dafc9016519e79184e383eab9ffbd579bebd/src/core/lombok/javac/handlers/JavacHandlerUtil.java#L845-L847

File: provisioning/provisioning-service-client/src/main/java/com/microsoft/azure/sdk/iot/provisioning/service/configs/DeviceRegistrationState.java (hunk 3 above)
Comment (abhipsaMisra) [reply provisioning/provisioning-service-client/src/main/java/com/microsoft/azure/sdk/iot/provisioning/service/configs/DeviceRegistrationState.java:66]: this sure was confusing to get the flow -> getLastUpdatedDateTimeUtc()  returns Date lastUpdatedDateTimeUtcDate which it constructed out of a string lastUpdatedDateTimeUtc

File: provisioning/provisioning-service-client/src/main/java/com/microsoft/azure/sdk/iot/provisioning/service/configs/EnrollmentGroup.java [hunk 7]
```diff
@@ -94,10 +96,22 @@
...
     private static final String ENROLLMENT_GROUP_ID_TAG = "enrollmentGroupId";
+
+    /**
+     * The unique identifier for this enrollment group.
+     *
+     * <p>
+     * A valid enrollment group Id shall follow this criteria.
+     * A case-sensitive string (up to 128 char long)
+     * of ASCII 7-bit alphanumeric chars
+     * + {'-', ':', '.', '+', '%', '_', '#', '*', '?', '!', '(', ')', ',', '=', '@', ';', '$', '''}.
```
Comment (abhipsaMisra) [inline provisioning/provisioning-service-client/src/main/java/com/microsoft/azure/sdk/iot/provisioning/service/configs/EnrollmentGroup.java:108]: how about we link to the ms doc instead, so that we don't have to update this if service changes their criteria? we don't do any client side validations, right?

File: provisioning/provisioning-service-client/src/main/java/com/microsoft/azure/sdk/iot/provisioning/service/configs/X509CertificateInfo.java [hunk 8]
```diff
@@ -42,50 +44,66 @@
...
+    @Getter
+    @Setter
     private String sha1Thumbprint;
 
     // the sha256 thumbprint of the X509 certificate
     private static final String SHA256_THUMBPRINT_TAG = "sha256Thumbprint";
     @Expose
     @SerializedName(SHA256_THUMBPRINT_TAG)
+    @Getter
+    @Setter
```
Comment (abhipsaMisra) [inline provisioning/provisioning-service-client/src/main/java/com/microsoft/azure/sdk/iot/provisioning/service/configs/X509CertificateInfo.java:64]: a couple of these need the @NotNull tag

File: provisioning/provisioning-service-client/src/test/java/com/microsoft/azure/sdk/iot/provisioning/service/configs/IndividualEnrollmentTest.java [hunk 9]
```diff
@@ -228,12 +228,12 @@ public void constructorWithJsonSetsOptionalParametersUsesSetters()
 
         // assert
         assertEquals(VALID_DEVICE_ID, Deencapsulation.getField(enrollment, "deviceId"));
-        assertNotNull(Deencapsulation.getField(enrollment, "registrationState"));
+        assertNotNull(enrollment.getDeviceRegistrationState());
```
Comment (abhipsaMisra) [inline provisioning/provisioning-service-client/src/test/java/com/microsoft/azure/sdk/iot/provisioning/service/configs/IndividualEnrollmentTest.java:231]: why was this change required?

File: provisioning/provisioning-service-client/src/test/java/com/microsoft/azure/sdk/iot/provisioning/service/configs/TpmAttestationTest.java (hunk 4 above)
Comment (abhipsaMisra) [reply provisioning/provisioning-service-client/src/test/java/com/microsoft/azure/sdk/iot/provisioning/service/configs/TpmAttestationTest.java:21]: could you add some context to why these null/empty checks were removed? is it because those values would only be set by service and hence would never be null/empty?

File: provisioning/provisioning-service-client/src/main/java/com/microsoft/azure/sdk/iot/provisioning/service/configs/DeviceRegistrationState.java (hunk 3 above)
Comment (timtay-microsoft) [reply provisioning/provisioning-service-client/src/main/java/com/microsoft/azure/sdk/iot/provisioning/service/configs/DeviceRegistrationState.java:66]: Yeah, the real unfortunate part of Lombok is that the getter's name is just get<FieldName> with no room for customization

File: provisioning/provisioning-service-client/src/test/java/com/microsoft/azure/sdk/iot/provisioning/service/configs/TpmAttestationTest.java (hunk 4 above)
Comment (timtay-microsoft) [reply provisioning/provisioning-service-client/src/test/java/com/microsoft/azure/sdk/iot/provisioning/service/configs/TpmAttestationTest.java:21]: It was a mix of fields where the service would validate them for us and readonly fields that the service wouldn't return null values for

File: provisioning/provisioning-service-client/src/main/java/com/microsoft/azure/sdk/iot/provisioning/service/Query.java (hunk 5 above)
Comment (timtay-microsoft) [reply provisioning/provisioning-service-client/src/main/java/com/microsoft/azure/sdk/iot/provisioning/service/Query.java:103]: This was a warning that I missed removing back when I was deleting deprecated methods. This method called into a deprecated method at one point, but no longer does, so we can safely remove this suppression

File: provisioning/provisioning-service-client/src/main/java/com/microsoft/azure/sdk/iot/provisioning/service/auth/ProvisioningConnectionString.java (hunk 6 above)
Comment (timtay-microsoft) [reply provisioning/provisioning-service-client/src/main/java/com/microsoft/azure/sdk/iot/provisioning/service/auth/ProvisioningConnectionString.java:31]: No reason in particular, but I can unify them to "PACKAGE" if that would make more sense

File: provisioning/provisioning-service-client/src/main/java/com/microsoft/azure/sdk/iot/provisioning/service/configs/EnrollmentGroup.java (hunk 7 above)
Comment (timtay-microsoft) [reply provisioning/provisioning-service-client/src/main/java/com/microsoft/azure/sdk/iot/provisioning/service/configs/EnrollmentGroup.java:108]: I like that idea, but this PR is really just moving existing documentation around, so I'll save editing it for later

File: provisioning/provisioning-service-client/src/test/java/com/microsoft/azure/sdk/iot/provisioning/service/configs/EnrollmentGroupTest.java [hunk 10]
```diff
@@ -1170,19 +1170,6 @@ public void setIotHubHostNameSucceed()
         assertEquals(newHostName, Deencapsulation.getField(enrollmentGroup, "iotHubHostName"));
     }
 
-    /* Tests_SRS_ENROLLMENT_GROUP_21_024: [The setInitialTwin shall throw IllegalArgumentException if the initialTwin is null.] */
-    @Test (expected = IllegalArgumentException.class)
-    public void setInitialTwinThrowsOnNull()
```
[the remark this answers is not in the document]
Comment (timtay-microsoft) [reply provisioning/provisioning-service-client/src/test/java/com/microsoft/azure/sdk/iot/provisioning/service/configs/EnrollmentGroupTest.java:1175]: The service does this validation for us, yeah

File: provisioning/provisioning-service-client/src/test/java/com/microsoft/azure/sdk/iot/provisioning/service/configs/IndividualEnrollmentTest.java (hunk 9 above)
Comment (timtay-microsoft) [reply provisioning/provisioning-service-client/src/test/java/com/microsoft/azure/sdk/iot/provisioning/service/configs/IndividualEnrollmentTest.java:231]: It wasn't required, but the method could be accessed directly rather than through reflection since the field is package private, not private. If we ever refactor the getDeviceRegistrationState() method, this unit test would get updated for free

File: provisioning/provisioning-service-client/src/main/java/com/microsoft/azure/sdk/iot/provisioning/service/configs/X509CertificateInfo.java (hunk 8 above)
Comment (timtay-microsoft) [reply provisioning/provisioning-service-client/src/main/java/com/microsoft/azure/sdk/iot/provisioning/service/configs/X509CertificateInfo.java:64]: The service will validate this info for us

File: provisioning/provisioning-service-client/src/main/java/com/microsoft/azure/sdk/iot/provisioning/service/auth/ProvisioningConnectionString.java (hunk 6 above)
Comment (abhipsaMisra) [reply provisioning/provisioning-service-client/src/main/java/com/microsoft/azure/sdk/iot/provisioning/service/auth/ProvisioningConnectionString.java:31]: I think it'll be good to use one throughout, esp since they fall into the same case. 
+1 to unifying them to "package", since it's closer to java's package package-private access level

(On a related note, I'd have thought Lombok AccessLevel.None would have been java-package-private (no modifier). Interesting that it is "private" instead).

**Review by abhipsaMisra — APPROVED** [review]

File: provisioning/provisioning-service-client/src/main/java/com/microsoft/azure/sdk/iot/provisioning/service/configs/BulkEnrollmentOperationError.java [hunk 11]
```diff
@@ -37,53 +38,23 @@
     private static final String REGISTRATION_ID_TAG = "registrationId";
     @Expose
     @SerializedName(REGISTRATION_ID_TAG)
+    @Getter
```
Comment (jamdavi) [inline provisioning/provisioning-service-client/src/main/java/com/microsoft/azure/sdk/iot/provisioning/service/configs/BulkEnrollmentOperationError.java:41]:
```suggestion
    /**
     * Getter for the error registrationId.
     *
     * @return The {@code String} with the registrationId content. It cannot be {@code null}.
     */
    @Getter
```

Just to note that with the @Getter methods you can also include full doc comments that will get generated when the lombok plugin executes for Maven. See

https://projectlombok.org/features/GetterSetter

File: provisioning/provisioning-service-client/src/main/java/com/microsoft/azure/sdk/iot/provisioning/service/configs/BulkEnrollmentOperationError.java [hunk 12]
```diff
@@ -37,53 +38,23 @@
...
     private String registrationId;
 
     // the error code
     private static final String ERROR_CODE_TAG = "errorCode";
     @Expose
     @SerializedName(ERROR_CODE_TAG)
+    @Getter
```
Comment (jamdavi) [inline provisioning/provisioning-service-client/src/main/java/com/microsoft/azure/sdk/iot/provisioning/service/configs/BulkEnrollmentOperationError.java:48]:
```suggestion
    /**
     * Getter for the error code.
     *
     * @return The {@code Integer} with the errorCode content. It cannot be {@code null}.
     */
     @Getter
```

File: provisioning/provisioning-service-client/src/main/java/com/microsoft/azure/sdk/iot/provisioning/service/configs/BulkEnrollmentOperationError.java [hunk 13]
```diff
@@ -37,53 +38,23 @@
...
     private Integer errorCode;
 
     // the error status
     private static final String ERROR_STATUS_TAG = "errorStatus";
     @Expose
     @SerializedName(ERROR_STATUS_TAG)
+    @Getter
```
Comment (jamdavi) [inline provisioning/provisioning-service-client/src/main/java/com/microsoft/azure/sdk/iot/provisioning/service/configs/BulkEnrollmentOperationError.java:55]:
```suggestion
    /**
     * Getter for the error status.
     *
     * @return The {@code String} with the errorStatus content. It can be {@code null}.
     */
     @Getter
```

**Review by jamdavi — APPROVED** [review]
Just a note about the @Getter/Setter code comments. 

You don't need to add them for ALL of the instances, but if we have additional context that would be lost it'd be good to get it in a code comment above the annotation. 

Lombok will generate correct code docs for us.

File: provisioning/provisioning-service-client/src/main/java/com/microsoft/azure/sdk/iot/provisioning/service/auth/ProvisioningConnectionString.java (hunk 6 above)
Comment (timtay-microsoft) [reply provisioning/provisioning-service-client/src/main/java/com/microsoft/azure/sdk/iot/provisioning/service/auth/ProvisioningConnectionString.java]: Sure thing, I'll mark all of these as PACKAGE for consistency
